hoomd/hpmc/external/field.py

```diff
@@ -1,243 +1,223 @@
+        V_{i}(q_i)  = \frac{1}{2} k_{rot} \cdot \min_j \left[ (q_i-q_{0,i} \cdot
+            q_{\mathrm{symmetry},j})^2 \right]
 
-    def set_params(self, k, q):
-        R""" Set the translational and rotational spring constants.
+    where :math:`k_{trans}` and :math:`k_{rot}` correspond to the arguments
+    ``k_translational`` and ``k_rotational``, respectively, :math:`r_i` and
+    :math:`q_i` are the position and orientation of particle :math:`i`, the
+    :math:`0` subscripts denote the given reference quantities, and
+    :math:`q_{\mathrm{symmetry}}` is the given set of symmetric orientations
+    from the ``symmetries`` argument.
 
-        Args:
-            k (float): translational spring constant.
-            q (float): rotational spring constant.
+    Attributes:
+        reference_positions (np.ndarray, shape=(*N_particles*, 3), dtype=`float`)
+            The reference positions, to which particles are restrained
+            :math:`[\mathrm{length}]`.
 
-        Example::
+        reference_orientations (np.ndarray, shape=(*N_particles*, 4), dtype=`float`)
+            The reference orientations, to which particles are
+            restrained :math:`[\mathrm{dimensionless}]`.
 
-            mc = hpmc.integrate.sphere(seed=415236);
-            lattice = hpmc.field.lattice_field(mc=mc, position=fcc_lattice, k=1000.0);
-            ks = np.linspace(1000, 0.01, 100);
-            for k in ks:
-              lattice.set_params(k=k, q=0.0);
-              run(1000)
+        k_translational (`float`): The translational spring constant
+            :math:`[\mathrm{energy} \cdot \mathrm{length}^{-2}]`.
 
-        """
-        self.cpp_compute.setParams(float(k), float(q))
-
-    def reset(self, timestep=None):
-        R""" Reset the statistics counters.
+        k_rotational (`float`): The rotational spring constant
+            :math:`[\mathrm{energy}]`.
 
-        Args:
-            timestep (int): the timestep to pass into the reset function.
+        symmetries (np.ndarray, shape=(*N_particles*, 4), dtype=`float`)
+            The orientations that are equivalent through symmetry,
+            i.e., the rotation quaternions that leave the particles unchanged
+            :math:`[\mathrm{dimensionless}]`.
 
-        Example::
+    """
 
-            mc = hpmc.integrate.sphere(seed=415236);
-            lattice = hpmc.field.lattice_field(mc=mc, position=fcc_lattice, k=1000.0);
-            ks = np.linspace(1000, 0.01, 100);
-            for k in ks:
-              lattice.set_params(k=k, q=0.0);
-              lattice.reset();
-              run(1000)
+    def __init__(self, reference_positions, reference_orientations,
+                 k_translational, k_rotational, symmetries):
+        param_dict = ParameterDict(
+            reference_positions=NDArrayValidator(dtype=np.float32,
+                                                 shape=(None, 3)),
+            reference_orientations=NDArrayValidator(dtype=np.float32,
```

Comment:
You require double's in the signatures in C++ but use single precision here. That involves a needless copy. Change one or the other.

Reply:
Changed.